atef/widgets/config/run.py

```diff
@@ -0,0 +1,285 @@
+""" helpers for run mode """
+
+import asyncio
+import logging
+from typing import Generator, List, Union
+
+from qtpy import QtCore
+from qtpy.QtWidgets import (QLabel, QLayout, QPushButton, QSpacerItem, QStyle,
+                            QToolButton, QWidget)
+
+from atef import util
+from atef.check import Result
+from atef.config import (AnyPreparedConfiguration, ConfigurationFile,
+                         PreparedComparison)
+from atef.enums import Severity
+from atef.procedure import ProcedureFile, ProcedureStep
+from atef.widgets.config.page import AtefItem, PageWidget
+from atef.widgets.core import DesignerDisplay
+
+logger = logging.getLogger(__name__)
+
+
+def walk_tree_items(item: AtefItem) -> Generator[AtefItem, None, None]:
+    yield item
+
+    for child_idx in range(item.childCount()):
+        yield from walk_tree_items(item.child(child_idx))
+
+
+FileType = Union[ProcedureFile, ConfigurationFile]
+
+
+def run_active_step(config):  # takes procedure steps and groups
+    """
+    Runs a given step and returns a result.
+    Does not change the verification status of the step, which is left to
+    other processes
+    """
+    result = config.run()
+
+    return result
+
+
+def make_run_page(
+    widget: QWidget,
+    configs: List[Union[PreparedComparison, ProcedureStep,
+                        AnyPreparedConfiguration]]
+) -> QWidget:
+    """
+    Disables all the widgets in ``widget`` and adds the RunCheck widget.
+    RunCheck widget holds the functionality needed to execute the
+    comparison or process
+
+    Add buttons to existing widget to ensure existing navigation methods
+    pass through
+
+    Parameters
+    ----------
+    widget : QWidget
+        widget to convert
+    configs : List[Union[PreparedComparison,
+                         ProcedureStep,
+                         AnyPreparedConfiguration]]
+        A list of atef dataclasses that can each hold a Result.
+
+    Returns
+    -------
+    QWidget
+        The run version of ``widget``
+    """
+    # TODO: consider options for different widget layouts
+    # currently assumes vertical layout.
+
+    # make existing widgets read-only
+    disable_widget(widget)
+    # add RunCheck to end of layout
+    check_widget = RunCheck(configs=configs)
+    widget.layout().addWidget(check_widget)
+    widget.run_check = check_widget
+
+    return widget
+
+
+def disable_widget(widget: QWidget) -> QWidget:
+    """ Disable widget, recurse through layouts """
```

Comment:
qt learning moment for me: what does this do that a simple `widget.setEnabled(False)` doesn't do? Is it possible to have a disabled parent widget with enabled children?

Reply:
Exactly.  For example, you might disable a parent widget, but the combo boxes or buttons inside a scroll area were still interactable

Reply:
That's wildly annoying behavior

Reply:
This may be something to review again in the future; maybe make a TODO?